consensus/src/dag/bootstrap.rs

```diff
@@ -159,42 +186,86 @@ impl TDagMode for SyncMode {
+        let handle = tokio::spawn(async move {
+            let result = sync_manager
+                .sync_dag_to(
+                    dag_fetcher,
+                    request,
+                    responders,
+                    sync_dag_store,
+                    commit_li,
+                )
+                .await;
+            let _ = res_tx.send(result);
+        });
+        defer!({
+            handle.abort();
+            let _ = block_on(handle);
+        });
+
+        let mut buffer = Vec::new();
+
+        select! {
+            biased;
+            res = res_rx => {
+                match res {
+                    Ok(sync_result) => {
+                        if let Ok(_) = sync_result {
+                            info!("sync succeeded. running full bootstrap.");
+                            // If the sync task finishes successfully, we can transition to Active mode by
+                            // rebootstrapping all components starting from the DAG store.
+                            let (new_state, new_handler, new_fetch_service) = bootstrapper.full_bootstrap();
+                            return Some(Mode::Active(ActiveMode {
+                                handler: new_handler,
+                                fetch_service: new_fetch_service,
+                                base_state: new_state,
+                                buffer,
+                            }))
+                        } else {
+                            info!("sync failed. resuming with current DAG state.");
+                            // If the sync task fails, then continue the DAG in Active Mode with existing state.
+                            let (new_handler, new_fetch_service) =
+                                bootstrapper.bootstrap_components(&self.base_state);
+                            return Some(Mode::Active(ActiveMode {
+                                handler: new_handler,
+                                fetch_service: new_fetch_service,
+                                base_state: self.base_state,
+                                buffer,
+                            }))
+                        }
+                    },
+                    Err(_) => unreachable!("sender won't be dropped without sending"),
+                }
+            },
+            res = network_handle.run(dag_rpc_rx, &mut buffer) => {
+                // The network handle returns if the sender side of dag_rpc_rx closes,
+                // or network handle found a future CertifiedNodeMessage to cancel the
+                // current sync.
+                if let Some(msg) = res {
+                    return Some(Mode::Sync(SyncMode {
+                        certified_node_msg: msg,
+                        base_state: self.base_state,
+                    }));
+                } else {
+                    unreachable!("remote mustn't drop the network message sender until bootstrapper returns");
```

Comment:
I don't think this is unreachable if there's an epoch change happening from epoch manager? but I guess it needs to get ack first before dropping the sender

Reply:
yes, that's my reasoning too.